### ISSUE TITLE
Macro-related terminology

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#11945]
#8835 sparked a sub-discussion about the terminology for certain macro features. In order to avoid further divergence from the original topic there, I'm extracting that to this issue.

Let's settle the terminology. There are several terms that are used for different things in some parts of the book and API docs, sometimes interchangeably, sometimes with different meanings.

1. *macro def*: Macros that are called from a macro expressions and return a macro expression, i.e. primitives from `Crystal::Macros` and user-defined as per #8835 (`macro def`).
2. *macro method*: Macros that are called from regular code and expand to code (`macro`).
3. *macro context* (?): Macro expressions in top-level code.
4. *???*: Method defs that use `@type` inside a macro expression and thus are instantiated once per type, not once per hierarchy (`def` + `{{ @type }}`)

I don't have a concise term for 4 and I'm not very convinced about 3. Alternative suggestions are welcome for everything.

Once we settle on clear terminology, we should comb through the docs to apply the agreed terms.

**Comment by beta-ziliani** [comment]
I'd call 3. "top-level macro", and I'm not sure we need a name for 4. It sounds more like a pattern than a thing on its own (or am I missing something?).

**Comment by Blacksmoke16** [comment]
Does 3 also include macro logic not at the top level, but also not within a method? E.g. a raw macro loop within a class used to generate methods. Having some term to reference using macro logic that isn't directly in a `macro method` would be good to have IMO.

Maybe the solution here is to do like:

* `macro` - General term to mean "a macro expression / method / def / context that happens within `{% %}` or `{{ }}`. Could be within a method, top level, within a namespace, etc.
* `macro def` - Same as OP
* `macro method` - Same as OP
* `macro context` - Just becomes `macro`
* Drop need for 4, as its just a method that has macro logic within it. I.e. can say "using @type within a method is instantiated once per type, not once per hierarchy (def + {{ @type }})" versus giving it a unique term.

**Comment by asterite** [comment]
I call those "inline macro"

**Comment by straight-shoota** [comment]
Yes, I mean not necessarily in top-level code but also macro code in regular methods.

**Comment by straight-shoota** [comment]
@beta-ziliani 4. is a special rule in the language. So I think it makes sense to call it something.

**Comment by beta-ziliani** [comment]
OK, but as @Blacksmoke16 said, I'm still not convinced it needs a name.

**Comment by HertzDevil** [comment]
IMO 4 is not so much a special rule as it is just one of the rules controlling def instantiation; argument types and argument positions also control def instantiations. If we are prepared to explain all of those in a dedicated section in the reference manual then not giving 4 a name is fine.
